File: Lesson9 Recursion and DP/Problem 4/Solutions/Solution.py
# submit this in pypy2, a faster version of python, but mainly this is python 2 code
# credits to dmoj user p1geon

import sys

def input():
    return sys.stdin.readline().strip()
    
c, m = map(int, input().split())
cages = [list(map(int, input().split())) for i in xrange(c)]
cost = [0]*(m + 1)
for i in xrange(1, c + 1):
    for j in reversed(xrange(cages[i - 1][0], m + 1)):
        cost[j] = max(cost[j], cages[i - 1][1] + cost[j - cages[i - 1][0]])
print cost[m]

File: Lesson9 Recursion and DP/Problem 4/Solutions/Solution.cpp
#include <iostream>
using namespace std;

int v[110], w[110]; long dp[100010];
int main(){
    int N, W;
    cin>>N>>W;
    for(int i = 1; i <= N; i++)
        cin>>w[i]>>v[i];

    for(int i = 1; i <= N; i++){
        for(int j = W; j >= w[i]; j--)
            dp[j] = max(dp[j], dp[j - w[i]] + v[i]);
    }
    cout<<dp[W]<<endl;
}
